lib/fog/libvirt/models/compute/server.rb

```diff
@@ -281,14 +283,28 @@ def to_xml
               end
 
               xml.vcpu(cpus)
-              xml.os do
+              os_tags = {}
+
+              # Set firmware only if it's EFI, BIOS don't need to be set
+              os_tags[:firmware] = "efi" if os_firmware == "efi"
+
+              xml.os(**os_tags) do
                 type = xml.type(os_type, :arch => arch)
                 type[:machine] = "q35" if ["i686", "x86_64"].include?(arch)
 
                 boot_order.each do |dev|
                   xml.boot(:dev => dev)
                 end
+
+                if os_firmware == "efi"
```

Comment:
Should this be:
```suggestion
                if os_firmware == "efi" && os_firmware_features.any?
```

Reply:
Also, please make sure we deal with `nil`
```suggestion
                if os_firmware == "efi" && os_firmware_features&.any?
```